### PR TITLE
Kernel_AccessValidator

### PR DESCRIPTION
**Pull request by mekk1t**
Latest update:

Added Unit Tests. **FINALLY**

### CODE REVIEW

File: DigitalOffice/Shared/Kernel/AccessValidator/Interfaces/IAccessValidator.cs
Comment (KostinP) [inline DigitalOffice/Shared/Kernel/AccessValidator/Interfaces/IAccessValidator.cs]: Add XML.

File: DigitalOffice/Shared/Kernel/AccessValidator/Requests/IAccessValidatorRequest.cs
Comment (KostinP) [inline DigitalOffice/Shared/Kernel/AccessValidator/Requests/IAccessValidatorRequest.cs]: Add XML.

File: DigitalOffice/Shared/Kernel/AccessValidator/Interfaces/IAccessValidator.cs
Comment (mekk1t) [reply DigitalOffice/Shared/Kernel/AccessValidator/Interfaces/IAccessValidator.cs]: Yeah, I'll refactor the code. It's the first draft, but thanks for the reminder.

File: DigitalOffice/Services/UserService/Broker/Consumers/AccessValidatorConsumer.cs
Comment (ifelseelif) [inline DigitalOffice/Services/UserService/Broker/Consumers/AccessValidatorConsumer.cs]: Here it is necessary to catch the exception from repository

Comment (mekk1t) [reply DigitalOffice/Services/UserService/Broker/Consumers/AccessValidatorConsumer.cs]: Good point. Will do

File: DigitalOffice/Shared/Kernel/AccessValidator/AccessValidator.cs
Comment (ifelseelif) [inline DigitalOffice/Shared/Kernel/AccessValidator/AccessValidator.cs]: Could you please explain how to send a userId via postman and receive userId here?

Comment (mekk1t) [reply DigitalOffice/Shared/Kernel/AccessValidator/AccessValidator.cs]: @ifelseelif well, I can't tell you how to test it via Postman, because I haven't done that yet.

I know for sure that this command works when you log in the web application via Identity and UserId must be stored in a cookie.

As for Postman, you can look up how to pass ASP Net Core authorization in Postman. I can't tell you much about the topic. I'm still waiting for the approval that I can implement Access Validator like this.

**Comment by mekk1t** [comment]
One note (if I forgot to write this in the commits):
Request-interface is kept in Kernel because there were compatibility issues and it was decided to leave the request model in the class library.

File: DigitalOffice/Shared/Kernel/AccessValidator/AccessValidator.cs
Comment (mekk1t) [reply DigitalOffice/Shared/Kernel/AccessValidator/AccessValidator.cs]: Implementation has changed to Headers. So if you'd want to test it via Postman, use them.

File: DigitalOffice/Services/UserService/Broker/Consumers/AccessValidatorConsumer.cs
Comment (mekk1t) [reply DigitalOffice/Services/UserService/Broker/Consumers/AccessValidatorConsumer.cs]: Done

File: DigitalOffice/Shared/Kernel/AccessValidator/Interfaces/IAccessValidator.cs
Comment (mekk1t) [reply DigitalOffice/Shared/Kernel/AccessValidator/Interfaces/IAccessValidator.cs]: Added

File: DigitalOffice/Shared/Kernel/AccessValidator/Requests/IAccessValidatorRequest.cs
Comment (mekk1t) [reply DigitalOffice/Shared/Kernel/AccessValidator/Requests/IAccessValidatorRequest.cs]: Added

File: DigitalOffice/Shared/Kernel/AccessValidator/Interfaces/IAccessValidator.cs
Comment (mekk1t) [reply DigitalOffice/Shared/Kernel/AccessValidator/Interfaces/IAccessValidator.cs]: Added

**Comment by mekk1t** [comment]
Checks are failing because of the incorrect mocks in AccessValidatorTests. I'm trying to fix it.

File: DigitalOffice/Shared/Kernel/AccessValidator/AccessValidator.cs
Comment (DarkFelas) [reply DigitalOffice/Shared/Kernel/AccessValidator/AccessValidator.cs]: image

**Review by DarkFelas — APPROVED** [review]

File: DigitalOffice/KernelUnitTests/AccessValidator/AccessValidatorTests.cs
Comment (lalagvana) [inline DigitalOffice/KernelUnitTests/AccessValidator/AccessValidatorTests.cs]: Extra spaces

**Review by lalagvana — APPROVED** [review]

File: Services/CheckRightsService/Repositories/Interfaces/ICheckRightsRepository.cs
Comment (KostinP) [inline Services/CheckRightsService/Repositories/Interfaces/ICheckRightsRepository.cs:19]: Add XML.

File: Services/ProjectService/Controllers/ProjectController.cs
Comment (KostinP) [inline Services/ProjectService/Controllers/ProjectController.cs:32]: :)

**Review by KostinP — APPROVED** [review]

File: KernelUnitTests/AccessValidator/AccessValidatorTests.cs
Comment (KostinP) [inline KernelUnitTests/AccessValidator/AccessValidatorTests.cs:63]: Extra empty lines

File: Shared/Kernel/AccessValidator/Interfaces/IAccessValidator.cs
Comment (KostinP) [inline Shared/Kernel/AccessValidator/Interfaces/IAccessValidator.cs:12]: Add XML.

File: Services/CheckRightsService/Database/Migrations/20200828092725_AddRightDbUser.cs
Comment (KostinP) [inline Services/CheckRightsService/Database/Migrations/20200828092725_AddRightDbUser.cs]: ???

File: KernelUnitTests/AccessValidator/AccessValidatorTests.cs
Comment (KostinP) [inline KernelUnitTests/AccessValidator/AccessValidatorTests.cs]: Extra empty lines.

**Review by KostinP — CHANGES_REQUESTED** [review]

File: Services/CheckRightsService/Repositories/Interfaces/ICheckRightsRepository.cs
Comment (KostinP) [inline Services/CheckRightsService/Repositories/Interfaces/ICheckRightsRepository.cs:25]: Add XML.

File: KernelUnitTests/AccessValidator/AccessValidatorTests.cs
Comment (djino97) [inline KernelUnitTests/AccessValidator/AccessValidatorTests.cs]: Can use OneTimeSetUp

File: Services/CheckRightsService/Broker/Consumers/AccessValidatorConsumer.cs
Comment (djino97) [inline Services/CheckRightsService/Broker/Consumers/AccessValidatorConsumer.cs]: Maybe here u can use OperationResultWrapper?

Comment (mekk1t) [reply Services/CheckRightsService/Broker/Consumers/AccessValidatorConsumer.cs]: Do I have to? I hate this method. I think this is much clearer to understand and to use.

Comment (djino97) [reply Services/CheckRightsService/Broker/Consumers/AccessValidatorConsumer.cs]: It violates the principles DRY, https://github.com/LT-Students/DigitalOffice/pull/54#discussion_r471047472

Comment (djino97) [reply Services/CheckRightsService/Broker/Consumers/AccessValidatorConsumer.cs]: I understand that someone may not like this construction, but if we decided to use it everywhere within the team, then it is better to do so

Comment (mekk1t) [reply Services/CheckRightsService/Broker/Consumers/AccessValidatorConsumer.cs]: meh. ok

File: Services/ProjectService/Controllers/ProjectController.cs
Comment (denis-kriv) [inline Services/ProjectService/Controllers/ProjectController.cs]: do we need this in develop?

File: Services/CheckRightsService/Broker/Consumers/AccessValidatorConsumer.cs
Comment (denis-kriv) [inline Services/CheckRightsService/Broker/Consumers/AccessValidatorConsumer.cs]: `var`

File: Services/CheckRightsService/Repositories/CheckRightsRepository.cs
Comment (denis-kriv) [inline Services/CheckRightsService/Repositories/CheckRightsRepository.cs:36]: why not to return false?

File: Services/UserService/Broker/Consumers/AccessValidatorConsumer.cs
Comment (denis-kriv) [inline Services/UserService/Broker/Consumers/AccessValidatorConsumer.cs]: `var`

File: Shared/Kernel/AccessValidator/AccessValidator.cs
Comment (denis-kriv) [inline Shared/Kernel/AccessValidator/AccessValidator.cs]: why do we use `_` in naming?

File: Services/CheckRightsService/Broker/Consumers/AccessValidatorConsumer.cs
Comment (denis-kriv) [inline Services/CheckRightsService/Broker/Consumers/AccessValidatorConsumer.cs]: `FromServices` ?

File: Services/UserService/Broker/Consumers/AccessValidatorConsumer.cs
Comment (denis-kriv) [inline Services/UserService/Broker/Consumers/AccessValidatorConsumer.cs]: `FromServices` ?

File: Services/ProjectService/Controllers/ProjectController.cs
Comment (mekk1t) [reply Services/ProjectService/Controllers/ProjectController.cs:44]: Jesus. Why the H did we start using XML? Who do I type this for? Or it's written in invisible ink?
image

Comment (mekk1t) [reply Services/ProjectService/Controllers/ProjectController.cs:44]: Or it's much easier to just address the question?

Comment (mekk1t) [reply Services/ProjectService/Controllers/ProjectController.cs]: Read the XML.
One typo though: It'll be deleted before PR merge, not after. But you get the idea

File: Services/UserService/Broker/Consumers/AccessValidatorConsumer.cs
Comment (mekk1t) [reply Services/UserService/Broker/Consumers/AccessValidatorConsumer.cs]: var...what?

File: Services/CheckRightsService/Database/Migrations/20200828092725_AddRightDbUser.cs
Comment (mekk1t) [reply Services/CheckRightsService/Database/Migrations/20200828092725_AddRightDbUser.cs]: So that dbContext query works properly.

File: Services/CheckRightsService/Repositories/Interfaces/ICheckRightsRepository.cs
Comment (mekk1t) [reply Services/CheckRightsService/Repositories/Interfaces/ICheckRightsRepository.cs:25]: Added

File: Services/CheckRightsService/Broker/Consumers/AccessValidatorConsumer.cs
Comment (mekk1t) [reply Services/CheckRightsService/Broker/Consumers/AccessValidatorConsumer.cs]: Actually, with a wrapper this looks much cleaner.
Although I have to admit, that this is not an "OperationResult" wrapper per se. It just creates an object.
Maybe wrapper can be modified to really wrap this into IOperationResult<T> -- now that would be cooler.

File: Services/CheckRightsService/Repositories/CheckRightsRepository.cs
Comment (mekk1t) [reply Services/CheckRightsService/Repositories/CheckRightsRepository.cs:36]: It's under discussion.

File: Services/ProjectService/Controllers/ProjectController.cs
Comment (mekk1t) [reply Services/ProjectService/Controllers/ProjectController.cs]: Anyway, I better delete that. But please don't ask any questions that are asked **one line above** the code you comment.

File: Services/CheckRightsService/Database/Migrations/20200828092725_AddRightDbUser.cs
Comment (mekk1t) [reply Services/CheckRightsService/Database/Migrations/20200828092725_AddRightDbUser.cs]: Anyway, I removed "delete data" sections.

File: Services/CheckRightsService/Database/Entities/DbRight.cs
Comment (KostinP) [inline Services/CheckRightsService/Database/Entities/DbRight.cs]: You can remove this code.

File: Services/ProjectService/Controllers/ProjectController.cs
Comment (KostinP) [inline Services/ProjectService/Controllers/ProjectController.cs]: Chek warnings
"The using directive for 'LT.DigitalOffice.ProjectService.Models' appeared previously in this namespace"

File: Services/ProjectService/Startup.cs
Comment (KostinP) [inline Services/ProjectService/Startup.cs]: The using directives appeared previously in this namespace

File: Shared/Kernel/AccessValidator/AccessValidator.cs
Comment (KostinP) [inline Shared/Kernel/AccessValidator/AccessValidator.cs]: Extra empty lines.

File: Tests/KernelTests/AccessValidator/AccessValidatorTests.cs
Comment (KostinP) [inline Tests/KernelTests/AccessValidator/AccessValidatorTests.cs]: Not "If", but "When".

File: Tests/KernelTests/AccessValidator/AccessValidatorTests.cs
Comment (KostinP) [inline Tests/KernelTests/AccessValidator/AccessValidatorTests.cs]: Not "If", but "When".

File: Tests/KernelTests/AccessValidator/AccessValidatorTests.cs
Comment (KostinP) [inline Tests/KernelTests/AccessValidator/AccessValidatorTests.cs]: Not "If", but "When".

File: Tests/KernelTests/AccessValidator/AccessValidatorTests.cs
Comment (KostinP) [inline Tests/KernelTests/AccessValidator/AccessValidatorTests.cs]: Not "If", but "When".

File: Tests/KernelTests/AccessValidator/AccessValidatorTests.cs
Comment (KostinP) [inline Tests/KernelTests/AccessValidator/AccessValidatorTests.cs]: Not "If", but "When".

File: Tests/KernelTests/AccessValidator/AccessValidatorTests.cs
Comment (KostinP) [inline Tests/KernelTests/AccessValidator/AccessValidatorTests.cs]: Not "If", but "When".

**Review by KostinP — CHANGES_REQUESTED** [review]

File: Services/CheckRightsService/Database/Entities/DbRight.cs
Comment (mekk1t) [reply Services/CheckRightsService/Database/Entities/DbRight.cs]: Why?

Comment (mekk1t) [reply Services/CheckRightsService/Database/Entities/DbRight.cs]: How else am I gonna get access to the related data?

Comment (mekk1t) [reply Services/CheckRightsService/Database/Entities/DbRight.cs]: Through direct DbSet<DbRightUser> UserRights?

File: Services/ProjectService/Controllers/ProjectController.cs
Comment (mekk1t) [reply Services/ProjectService/Controllers/ProjectController.cs]: Fixed

File: Services/ProjectService/Startup.cs
Comment (mekk1t) [reply Services/ProjectService/Startup.cs]: Fixed

File: Shared/Kernel/AccessValidator/AccessValidator.cs
Comment (mekk1t) [reply Shared/Kernel/AccessValidator/AccessValidator.cs]: Fixed

File: Tests/KernelTests/AccessValidator/AccessValidatorTests.cs
Comment (mekk1t) [reply Tests/KernelTests/AccessValidator/AccessValidatorTests.cs]: Fixed

File: Tests/KernelTests/AccessValidator/AccessValidatorTests.cs
Comment (mekk1t) [reply Tests/KernelTests/AccessValidator/AccessValidatorTests.cs]: Fixed

File: Tests/KernelTests/AccessValidator/AccessValidatorTests.cs
Comment (mekk1t) [reply Tests/KernelTests/AccessValidator/AccessValidatorTests.cs]: Fixed

File: Tests/KernelTests/AccessValidator/AccessValidatorTests.cs
Comment (mekk1t) [reply Tests/KernelTests/AccessValidator/AccessValidatorTests.cs]: Fixed

File: Tests/KernelTests/AccessValidator/AccessValidatorTests.cs
Comment (mekk1t) [reply Tests/KernelTests/AccessValidator/AccessValidatorTests.cs]: Fixed

File: Tests/KernelTests/AccessValidator/AccessValidatorTests.cs
Comment (mekk1t) [reply Tests/KernelTests/AccessValidator/AccessValidatorTests.cs]: Fixed

File: Services/CheckRightsService/Database/Entities/DbRight.cs
Comment (KostinP) [reply Services/CheckRightsService/Database/Entities/DbRight.cs]: DbRightConfiguration is empty

Comment (mekk1t) [reply Services/CheckRightsService/Database/Entities/DbRight.cs]: Fixed

File: Services/CheckRightsService/Repositories/CheckRightsRepository.cs
Comment (KostinP) [inline Services/CheckRightsService/Repositories/CheckRightsRepository.cs]: Fix spaces.

File: Services/CompanyService/appsettings.json
Comment (KostinP) [inline Services/CompanyService/appsettings.json]: Most likely this is true, but can you explain why this is so?

File: Services/ProjectService/Startup.cs
Comment (KostinP) [inline Services/ProjectService/Startup.cs:72]: Fix spaces.

File: Services/ProjectService/appsettings.json
Comment (KostinP) [inline Services/ProjectService/appsettings.json:19]: Fix spaces.

File: Tests/KernelTests/AccessValidator/AccessValidatorTests.cs
Comment (KostinP) [inline Tests/KernelTests/AccessValidator/AccessValidatorTests.cs]: Allows to pass harness. Bus into the AccessValidator constructor. Otherwise it is null.

**Review by KostinP — CHANGES_REQUESTED** [review]
After the next edit, I will check the tests.

File: Shared/Kernel/AccessValidator/AccessValidator.cs
Comment (djino97) [inline Shared/Kernel/AccessValidator/AccessValidator.cs]: Add emty line

File: Tests/KernelTests/AccessValidator/Consumers/CheckRightsServiceConsumerTests.cs
Comment (djino97) [inline Tests/KernelTests/AccessValidator/Consumers/CheckRightsServiceConsumerTests.cs]: Can be separated declaration Mock variable from usual

**Review by djino97 — APPROVED** [review]

File: Services/CheckRightsService/Repositories/CheckRightsRepository.cs
Comment (mekk1t) [reply Services/CheckRightsService/Repositories/CheckRightsRepository.cs]: Jesus you won't let me close this will ya.
Should this be fixed? It's crystal clear to read, @lalagvana  liked it, I like it.
This is subjective. Anyway we agreed to indent LINQ methods, didn't we?

File: Services/CompanyService/appsettings.json
Comment (mekk1t) [reply Services/CompanyService/appsettings.json]: What do you mean exactly? 
I'm negotiating with Spartak what our appsettings should look like. I'll commit the approved version. This can be ignored

File: Tests/KernelTests/AccessValidator/AccessValidatorTests.cs
Comment (mekk1t) [reply Tests/KernelTests/AccessValidator/AccessValidatorTests.cs]: Um.... what? .Bus is the property. Stop nit-picking my PR.

File: Tests/KernelTests/AccessValidator/Consumers/CheckRightsServiceConsumerTests.cs
Comment (mekk1t) [reply Tests/KernelTests/AccessValidator/Consumers/CheckRightsServiceConsumerTests.cs]: Will fix

File: Services/ProjectService/Startup.cs
Comment (mekk1t) [reply Services/ProjectService/Startup.cs:72]: What should I fix? Is this really important? jesus
image

File: Services/ProjectService/appsettings.json
Comment (mekk1t) [reply Services/ProjectService/appsettings.json:19]: And again I don't see _any_ reason to fix _anything_.
image

File: Tests/KernelTests/AccessValidator/Consumers/CheckRightsServiceConsumerTests.cs
Comment (mekk1t) [reply Tests/KernelTests/AccessValidator/Consumers/CheckRightsServiceConsumerTests.cs]: Fixed

File: Shared/Kernel/AccessValidator/AccessValidator.cs
Comment (mekk1t) [reply Shared/Kernel/AccessValidator/AccessValidator.cs]: Done

File: Tests/KernelTests/AccessValidator/AccessValidatorTests.cs
Comment (mekk1t) [reply Tests/KernelTests/AccessValidator/AccessValidatorTests.cs]: Ah, capitalize the first letter. Well that's some big deal problems here alright

Comment (mekk1t) [reply Tests/KernelTests/AccessValidator/AccessValidatorTests.cs]: Fixed the message.

File: Services/CompanyService/appsettings.json
Comment (mekk1t) [reply Services/CompanyService/appsettings.json]: But i think this structure should be used everywhere. It provides a clearer configuration and you need to make changes only in the appsettings.json file.
Also all services that will be calling AccessValidator have to have this section.

File: Services/CheckRightsService/Repositories/CheckRightsRepository.cs
Comment (mekk1t) [reply Services/CheckRightsService/Repositories/CheckRightsRepository.cs]: Enjoy. Hope you like this version much more.

**Review by KostinP — APPROVED** [review]

**Pull request merged**
